go/host/l1/dataservice.go

```diff
@@ -467,6 +466,45 @@ func (r *DataService) streamLiveBlocks() {
 	}
 }
 
+func (r *DataService) fetchAncestors(blockHash gethcommon.Hash) error {
+	_, err := r.blockResolver.ReadBlock(&blockHash)
+	if err == nil {
+		return nil
+	}
+
+	// fetch the missing block from the l1 rpc
+	block, err := r.ethClient.HeaderByHash(blockHash)
+	if err != nil {
+		return fmt.Errorf("error fetching block: %w", err)
+	}
+
+	if block.Number.Uint64() <= common.L1GenesisHeight+1 {
+		return nil
+	}
+	// recursively check that the ancestors
+	err = r.fetchAncestors(block.ParentHash)
+	if err != nil {
+		return fmt.Errorf("error fetching ancestor: %w", err)
+	}
+
+	// handle the blocks in the reverse order of the stack - from the oldest to the newest
+	r.handleNewBlock(block)
```

Comment:
I don't really like the idea of notifying subscribers of a 'new head' when you're searching backwards here. The guardians use that newHead notification to decide if they are in catch-up mode or not, which could be dangerous for the sequencer, if it thinks it is up-to-date it will publish batches linked to old L1 blocks.

I don't mind the idea of backfilling the db but I wouldn't notify subscribers of new block I don't think.